File: p933e/recent_counter.py
from collections import deque


class RecentCounter:
    def __init__(self):
        self.queue = deque()

    def ping(self, t: int) -> int:
        self.queue.append(t)
        while len(self.queue) > 0 and t - self.queue[0] > 3000:
            self.queue.popleft()
        return len(self.queue)


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)
tests = [
    ([1], [1]),
    ([1, 100, 3001, 3002], [1, 2, 3, 3]),
    ([1, 100, 5001, 5002, 8300], [1, 2, 1, 2, 1]),
]
for times, expected in tests:
    rc = RecentCounter()
    actual = [rc.ping(t) for t in times]
    assert actual == expected

File: p933e/recent_counter.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <queue>
#include <iostream>
#include <cassert>

using namespace std;

class RecentCounter {
    queue<int> calls;
public:
    RecentCounter():
    calls() { }
    
    int ping(int t) {
        calls.push(t);
        while (!calls.empty() && t - calls.front() > 3000) {
            calls.pop();
        }
        return calls.size();
    }
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */
int main(int argc, char* argv[]) {
    RecentCounter obj;
    assert(obj.ping(1) == 1);
    assert(obj.ping(100) == 2);
    assert(obj.ping(3001) == 3);
    assert(obj.ping(3002) == 3);
    return 0;
}
